src/components/molecules/OakModal/OakModal.tsx

```diff
@@ -0,0 +1,196 @@
+import React, {
+  HTMLAttributes,
+  ReactNode,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from "react";
+import { createPortal } from "react-dom";
+import { FocusOn } from "react-focus-on";
+import { Transition, TransitionStatus } from "react-transition-group";
+import styled from "styled-components";
+
+import { InternalShadowRoundButton } from "../InternalShadowRoundButton";
+
+import { OakBox, OakFlex, OakImage } from "@/components/atoms";
+import { parseOpacity } from "@/styles/helpers/parseOpacity";
+
+export type OakModalProps = {
+  /**
+   * The content of the modal.
+   * Use with `<OakModalBody>` for best results.
+   */
+  children: ReactNode;
+  /**
+   * Slot for the footer of the modal.
+   * Use with `<OakModalFooter>` for best results.
+   */
+  footerSlot?: ReactNode;
+  /**
+   * Indicates whether the modal is open or closed
+   */
+  isOpen: boolean;
+  /**
+   * Called when the modal is closed
+   */
+  onClose: () => void;
+  /**
+   * The DOM container to render the modal portal into.
+   *
+   * @default document.body
+   */
+  domContainer?: Element;
+} & Pick<
+  HTMLAttributes<Element>,
+  "aria-label" | "aria-description" | "aria-labelledby" | "aria-describedby"
+>;
+
+const FadeOutBox = styled(OakBox)<{ $state: TransitionStatus }>`
+  opacity: ${({ $state }) => {
+    switch ($state) {
+      case "entered":
+      case "entering":
+        return parseOpacity("semi-transparent");
+      default:
+        return parseOpacity("transparent");
+    }
+  }};
+`;
+
+const SlideInFlex = styled(OakFlex)<{ $state: TransitionStatus }>`
+  transform: ${({ $state }) => {
+    switch ($state) {
+      case "entered":
+      case "entering":
+        return "translateX(0)";
+      default:
+        return "translateX(-100%)";
+    }
+  }};
+`;
+
+const logoSrc = `https://${process.env.NEXT_PUBLIC_OAK_ASSETS_HOST}/${process.env.NEXT_PUBLIC_OAK_ASSETS_PATH}/logo-mark.svg`;
+
+/**
+ * Modal dialog with trapped focus and a close button.
+ */
+export const OakModal = ({
+  children,
+  footerSlot,
+  domContainer = document.body,
+  isOpen,
+  onClose,
+  ...rest
+}: OakModalProps) => {
+  const [canaryElement, setCanaryElement] = useState<HTMLDivElement | null>(
```

Comment:
So the canaryElement is set to an empty div which can be used to attach the IntersectionObserver. This is inturn is used to update isScrolled state  ?

Reply:
that is correct. The canary sits at the top of the element and when it scrolls out of/into view (I.e it no longer / now intersects the viewport) the observer fires and we can use that to cheaply detect that the element has been scrolled.